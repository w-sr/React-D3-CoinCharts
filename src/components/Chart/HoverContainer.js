import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

import Flex from "../Flex";
import { PROPTYPES } from "../../constants";
import { animation, border, color, fontSize, fontWeight } from "../../styles/constants";

const HOVER_CONTAINER_WIDTH = 200;
export const VERTICAL_OFFSET = -12;

const StyledHoverContainer = styled(Flex).attrs({
  style: ({ left, position, visible }) => ({
    bottom: position === "bottom" ? `${VERTICAL_OFFSET}px` : undefined,
    top: position === "top" ? `${VERTICAL_OFFSET}px` : undefined,
    left: `${left - HOVER_CONTAINER_WIDTH / 2}px`,
    opacity: visible ? 1 : 0,
  }),
})`
  position: absolute;
  transition: opacity ${animation.speed};
  width: ${`${HOVER_CONTAINER_WIDTH}px`};
  z-index: 10;
`;

const Label = styled.div.attrs({
  background: props => (props.invertColor ? color.coinchartsGray : color.white),
  border: props => (props.invertColor ? "none" : border.border),
  color: props => (props.invertColor ? color.white : color.coinchartsGray),
})`
  background: ${props => props.background};
  border-radius: ${border.radius};
  border: ${props => props.border};
  color: ${props => props.color};
  font-size: ${fontSize.small};
  font-weight: ${fontWeight.medium}
  padding: 1px 6px;;
`;

const HoverContainer = ({ position, label, visible, x }) => (
  <StyledHoverContainer data-testid="HoverContainer" justify="center" left={x} position={position} visible={visible}>
    <Label invertColor={position === "top"}>{label}</Label>
  </StyledHoverContainer>
);

HoverContainer.propTypes = {
  position: PROPTYPES.HOVER_CONTAINER_POSITION.isRequired,
  visible: PropTypes.bool.isRequired,
  x: PropTypes.number.isRequired,
  label: PropTypes.string,
};

HoverContainer.defaultProps = {
  label: "",
};

export default HoverContainer;
