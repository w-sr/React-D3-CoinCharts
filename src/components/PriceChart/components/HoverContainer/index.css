.HoverContainer {
  font-size: 14px;
  font-weight: 500;
  display: flex;
  justify-content: center;
  position: absolute;
  width: 200px;
}

.HoverContainer.hidden {
  opacity: 0;
  transition: opacity 300ms;
}

.HoverContainer.show {
  opacity: 1;
  transition: opacity 300ms;
}

.HoverContainer .content {
  border-radius: 5px;
  padding: 1px 6px;
  background: #fff;
  border: 1px solid #7d95b6;
  color: #7d95b6;
}

.HoverContainer .content.invertColor {
  background: #7d95b6;
  border: none;
  color: #fff;
}
