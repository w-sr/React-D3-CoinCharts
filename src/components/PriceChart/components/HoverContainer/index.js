import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './index.css';

const HOVER_CONTAINER_WIDTH = 200;
const VERTICAL_OFFSET = -12;

const HoverContainer = ({ bottom, top, value, visible, x }) => {
  const containerClass = classNames({
    HoverContainer: true,
    hidden: !visible,
    show: visible,
  });

  const contentClass = classNames({
    content: true,
    invertColor: top,
  });

  return (
    <div
      className={containerClass}
      style={{
        bottom: bottom && VERTICAL_OFFSET,
        left: x - (HOVER_CONTAINER_WIDTH / 2),
        top: top && VERTICAL_OFFSET,
      }}
    >
      <div className={contentClass}>{value}</div>
    </div>
  );
};

HoverContainer.propTypes = {
  bottom: PropTypes.bool,
  top: PropTypes.bool,
  value: PropTypes.string,
  visible: PropTypes.bool,
  x: PropTypes.number,
};

HoverContainer.defaultProps = {
  bottom: false,
  top: false,
  value: '',
  visible: false,
  x: 0,
};

export default HoverContainer;
