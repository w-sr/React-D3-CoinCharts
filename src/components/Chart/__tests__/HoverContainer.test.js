import React from "react";
import { render } from "react-testing-library";

import HoverContainer, { VERTICAL_OFFSET } from "../HoverContainer";

const TEST_PROPS = {
  position: "top",
  visible: true,
  x: 0,
  label: "test-label",
};

describe("<HoverContainer />", () => {
  it("renders without crashing", () => {
    const { container } = render(<HoverContainer {...TEST_PROPS} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders `prop.label` correctly inside <HoverContainer/>", () => {
    const { container } = render(<HoverContainer {...TEST_PROPS} />);
    expect(container.firstChild.textContent).toEqual(TEST_PROPS.label);
  });

  it("shows component when `props.visible` is true", () => {
    const { container } = render(<HoverContainer {...TEST_PROPS} visible />);
    expect(container.firstChild.style.opacity).toEqual("1");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("hides component when `props.visible` is false", () => {
    const { container } = render(<HoverContainer {...TEST_PROPS} visible={false} />);
    expect(container.firstChild.style.opacity).toEqual("0");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("has `style.top` defined when `prop.position` is `top`", () => {
    const { container } = render(<HoverContainer {...TEST_PROPS} position="top" />);
    expect(container.firstChild.style.top).toEqual(`${VERTICAL_OFFSET}px`);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("has `style.bottom` defined when `prop.position` is `bottom`", () => {
    const { container } = render(<HoverContainer {...TEST_PROPS} position="bottom" />);
    expect(container.firstChild.style.bottom).toEqual(`${VERTICAL_OFFSET}px`);
    expect(container.firstChild).toMatchSnapshot();
  });
});
