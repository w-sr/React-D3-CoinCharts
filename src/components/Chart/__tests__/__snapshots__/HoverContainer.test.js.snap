// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HoverContainer /> has \`style.bottom\` defined when \`prop.position\` is \`bottom\` 1`] = `
<div
  class="sc-bwzfXH ogNWx sc-bdVaJa epQZNM"
  style="bottom: -12px; left: -100px; opacity: 1;"
  data-testid="HoverContainer"
>
  <div
    background="#FFFFFF"
    border="1px solid #DAE1E9"
    color="#7D95B6"
    class="sc-htpNat jaGnAe"
  >
    test-label
  </div>
</div>
`;

exports[`<HoverContainer /> has \`style.top\` defined when \`prop.position\` is \`top\` 1`] = `
<div
  class="sc-bwzfXH ogNWx sc-bdVaJa epQZNM"
  style="top: -12px; left: -100px; opacity: 1;"
  data-testid="HoverContainer"
>
  <div
    background="#7D95B6"
    border="none"
    color="#FFFFFF"
    class="sc-htpNat cufbqD"
  >
    test-label
  </div>
</div>
`;

exports[`<HoverContainer /> hides component when \`props.visible\` is false 1`] = `
<div
  class="sc-bwzfXH ogNWx sc-bdVaJa epQZNM"
  style="top: -12px; left: -100px; opacity: 0;"
  data-testid="HoverContainer"
>
  <div
    background="#7D95B6"
    border="none"
    color="#FFFFFF"
    class="sc-htpNat cufbqD"
  >
    test-label
  </div>
</div>
`;

exports[`<HoverContainer /> renders without crashing 1`] = `
<div
  class="sc-bwzfXH ogNWx sc-bdVaJa epQZNM"
  style="top: -12px; left: -100px; opacity: 1;"
  data-testid="HoverContainer"
>
  <div
    background="#7D95B6"
    border="none"
    color="#FFFFFF"
    class="sc-htpNat cufbqD"
  >
    test-label
  </div>
</div>
`;

exports[`<HoverContainer /> shows component when \`props.visible\` is true 1`] = `
<div
  class="sc-bwzfXH ogNWx sc-bdVaJa epQZNM"
  style="top: -12px; left: -100px; opacity: 1;"
  data-testid="HoverContainer"
>
  <div
    background="#7D95B6"
    border="none"
    color="#FFFFFF"
    class="sc-htpNat cufbqD"
  >
    test-label
  </div>
</div>
`;
